Microsoft.Toolkit.Mvvm/ComponentModel/Attributes/ObservablePropertyAttribute.cs

```diff
@@ -0,0 +1,44 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#pragma warning disable CS1574
+
+using System;
+using System.ComponentModel;
+
+namespace Microsoft.Toolkit.Mvvm.ComponentModel
+{
+    /// <summary>
+    /// An attribute that indicates that a given field should be wrapped by a generated observable property.
+    /// In order to use this attribute, the containing type has to implement the <see cref="INotifyPropertyChanged"/> interface
+    /// and expose a method with the same signature as <see cref="ObservableObject.OnPropertyChanged(string?)"/>. If the containing
+    /// type also implements the <see cref="INotifyPropertyChanging"/> interface and exposes a method with the same signature as
+    /// <see cref="ObservableObject.OnPropertyChanging(string?)"/>, then this method will be invoked as well by the property setter.
+    /// <para>
+    /// This attribute can be used as follows:
+    /// <code>
+    /// partial class MyViewModel : ObservableObject
+    /// {
+    ///     [ObservableProperty]
+    ///     private string name;
+    /// }
+    /// </code>
+    /// </para>
+    /// And with this, code analogous to this will be generated:
+    /// <code>
+    /// partial class MyViewModel
+    /// {
+    ///     public string Name
+    ///     {
+    ///         get => name;
+    ///         set => SetProperty(ref name, value);
+    ///     }
+    /// }
+    /// </code>
+    /// </summary>
+    [AttributeUsage(AttributeTargets.Field, AllowMultiple = false, Inherited = false)]
+    public sealed class ObservablePropertyAttribute : Attribute
+    {
```

Comment:
Because we have the problem above, do we want to have an alternate optional property/constructor for this which allows you to provide a custom alias for the name of the resulting property. e.g.

```cs
[ObservableProperty("IsEnabled")]
private bool isenabled
````

Reply:
See answer above. I don't think this is necessary given that 🤔

Reply:
What about if you want to upgrade existing code where the field name you're using to drive a property is named differently than the property you exposed?

```cs
    private int hitCounter;

   public int Count
   {
       get => hitCounter;
       set => SetProperty(ref hitCounter, value);
   }
```

Though I know they could just rename `hitCounter`, but maybe they have other internal counters and are distinguishing...

Reply:
I would expect people to just rename the private field here, yeah. I mean this would just be:

```csharp
[ObservableProperty]
private int count;
```

For other cases (which would be weird though) I'd expect users to just fallback into explicit setters for the property as usual. After all we're not trying to support every possible case ever (we couldn't anyway). I just really don't like having actual symbol names as a hardcoded string in an attribute, that feels really "non C#-ish" to me 😄

Reply:
Sure, just wanted to call this out as a potential thing. It's always something we can add later since the default is to just use the provided field name. So no harm in leaving for now and getting more feedback at large once we ship it.